docs: add finetuner docs (#771)
hanxiao/bert-as-service: docs/user-guides/finetuner.md

> # Fine-tune Models

Reviewer: Can we have a colab notebook for this purpose?

Author: i could, but have to add some loop to wait for finetuner.run

Reviewer: In practice, the user should also add a loop to make sure the job is done. This notebook can provide more practical tips then.

> # Fine-tuning CLIP Models Using Finetuner

Reviewer 2: i would make it a bit shorter, see 
![image](https://user-images.githubusercontent.com/9794489/179534219-46c76f77-51d7-4d69-bdee-e2a983f219e2.png)

Reviewer: use `Fine-tune a pretrained model` as the title? @ZiniuYu you can also borrow some idea from this doc https://huggingface.co/docs/transformers/training

Author: I will go with Fine-tune Models

> Although CLIP-as-service has provided you a list of pre-trained models, you can also train your own models. 

Reviewer 2: your own models -> your models

This guide will show you how to use [Finetuner](https://finetuner.jina.ai) to fine-tune models and use them in CLIP-as-service.

> For installation and basic usage of Finetuner, please refer to [Finetuner website](https://finetuner.jina.ai).

Reviewer: please redirect to the clip model finetuning tutorial link

Author: is this really needed? i think the main page is sufficient, from which users can find useful information. Otherwise we provide a different narrative of the same thing

Reviewer: But we need to guild the user to access the clip fineturning tutorial. So far, I cannot find this link in the doc.

## Prepare Training Data

Finetuner accepts training data and evaluation data in the form of [`DocumentArray`](https://docarray.jina.ai/fundamentals/documentarray/).

> The training data for CLIP is a list of (text, image) pairs.

Reviewer: Can we add a data sample here? I feel it is not easy to understand what's the dataset should looks like. I belive a real data sample could help the reader.

Author: let me think of an elegant way

Each pair is store in a [`Document`](https://docarray.jina.ai/fundamentals/document/) which wraps two [`chunks`](https://docarray.jina.ai/fundamentals/document/nested/) with the `image` and `text` modality.
You can push the resulting [`DocumentArray`](https://docarray.jina.ai/fundamentals/documentarray/) to the cloud using the [`.push`](https://docarray.jina.ai/api/docarray.array.document/?highlight=push#docarray.array.document.DocumentArray.push) method.
A sample to construct and push the training data is shown below.

```python
from docarray import Document, DocumentArray

train_da = DocumentArray(
    [
        Document(
            chunks=[
                Document(
                    content='pencil skirt slim fit available for sell',
                    modality='text',
                ),
                Document(
                    uri='https://...skirt-1.png',
                    modality='image',
                ),
            ],
        ),
    ]
)
train_da.push('name-of-train-data')
```

## Start Finetuner

You may now create and run a fine-tuning job after login to Jina ecosystem.

> For simplicity, we directly use `clip-fashion-train-data` and `clip-fashion-eval-data` provided by Finetuner.

Reviewer: The pre-defined dataset `clip-fashion` is an example dataset. Not **For simplicity**. This sentence will mislead the reader.

)
```

After the job started, you may use [`.status`](https://finetuner.jina.ai/api/finetuner.run/#finetuner.run.Run.status) to check the status of the job.

```python
import finetuner

finetuner.login()
run = finetuner.get_run('clip-fashion')
print(run.status())
```

When the status is `FINISHED`, you can download the tuned model to your local machine.

```python
import finetuner

finetuner.login()
run = finetuner.get_run('clip-fashion')
run.save_artifact('clip-model')
```

You should now get a zip file containing the tuned model named `clip-fashion.zip` under the folder `clip-model`.

## Use the Model

After unzipping the model you get from the previous step, a folder with the following structure will be generated:

```text
.
└── clip-fashion/
    ├── config.yml
    ├── metadata.yml
    ├── metrics.yml
    └── models/
        ├── clip-text/
        │   ├── metadata.yml
        │   └── model.onnx
        ├── clip-vision/
        │   ├── metadata.yml
        │   └── model.onnx
        └── input-map.yml
```

Since the tuned model generated from Finetuner contains richer information such as metadata and config, we now transform it to simpler structure used by CLIP-as-service.

> First create a new folder named `clip-fashion-cas` or anything you like. This will be the storage of the models to use in CLIP-as-service.

Reviewer: use a bulleted list to organize the instructions.

Then copy and move `clip-fashion/models/clip-text/model.onnx` to `clip-fashion-cas` and rename it to `textual.onnx`.

Similarly, copy and move `clip-fashion/models/clip-vision/model.onnx` to `clip-fashion-cas` and rename it to `visual.onnx`.

Now that you should have your clip-fashion-cas structured like this:

```text
.
└── clip-fashion-cas/
    ├── textual.onnx
    └── visual.onnx
```

> In order to use finetuned model, create a custom yaml file `finetuned_clip.yml`. For more information on flow and `clip_server` customization, please refer to [https://docs.jina.ai/fundamentals/flow/yaml-spec/](https://docs.jina.ai/fundamentals/flow/yaml-spec/) and [https://clip-as-service.jina.ai/user-guides/server/#yaml-config](https://clip-as-service.jina.ai/user-guides/server/#yaml-config)

Reviewer 2: these two links is not formated:

![image](https://user-images.githubusercontent.com/9794489/179533945-7bfb6650-9a7b-4b15-ba3c-7f307d05e784.png)

```yaml
jtype: Flow
version: '1'
with:
  port: 51000
executors:
  - name: clip_o
    uses:
      jtype: CLIPEncoder
      metas:
        py_modules:

>           - executors/clip_onnx.py

Reviewer: use the latest syntax for `py_modules`.

Author: compatibility issue?

Reviewer: We can bump the jina version.

      with:

>         name: ViT-B/32 # since finetuner only support ViT-B/32 for CLIP

Reviewer: polish the comment (I suggest adding a warning phrase to emphasize this part). Remind the user that the model name should match the fine-tuned model. Otherwise, the output is not correct.

Author: warning make sense

        model_path: 'clip-fashion-cas' # path to clip-fashion-cas
    replicas: 1
```

You can now start the `clip_server` using fine-tuned model to get a performance boost:

```bash
python -m clip_server finetuned_clip.yml
```

> That's it, enjoy!

Reviewer: Add a emoji 🚀

Author: :smile_wechat:

# Fine-tune Models

Although CLIP-as-service has provided you a list of pre-trained models, you can also fine-tune your models. 
This guide will show you how to use [Finetuner](https://finetuner.jina.ai) to fine-tune models and use them in CLIP-as-service.

For installation and basic usage of Finetuner, please refer to [Finetuner documentation](https://finetuner.jina.ai).
You can also [learn more details about fine-tuning CLIP](https://finetuner.jina.ai/tasks/text-to-image/).

## Prepare Training Data

> Finetuner accepts training data and evaluation data in the form of [`DocumentArray`](https://docarray.jina.ai/fundamentals/documentarray/).

Reviewer 3: Instead of having hyperlinks for these classes and methods from other repos, its quite nice to use the [intersphinx](https://www.sphinx-doc.org/en/master/usage/extensions/intersphinx.html) extension of sphinx. This line would then look a bit neater as follows:

```suggestion
Finetuner accepts training data and evaluation data in the form of {class}`~docarray.DocumentArray`.
```

You just need to specify in your conf.py extensions to include intersphinx and add your mentioned repos:
```
# intersphinx references
intersphinx_mapping = {'docarray': ('https://docarray.jina.ai/', None)}
```

Author: Ah yes, thanks for mentioning

# Fine-tune Models

Although CLIP-as-service has provided you a list of pre-trained models, you can also train your models. 
This guide will show you how to use [Finetuner](https://finetuner.jina.ai) to fine-tune models and use them in CLIP-as-service.

For installation and basic usage of Finetuner, please refer to [Finetuner documentation](https://finetuner.jina.ai).
You can also [learn more details about fine-tuning CLIP](https://finetuner.jina.ai/tasks/text-to-image/).

## Prepare Training Data

Finetuner accepts training data and evaluation data in the form of [`DocumentArray`](https://docarray.jina.ai/fundamentals/documentarray/).
The training data for CLIP is a list of (text, image) pairs.

> Each pair is store in a [`Document`](https://docarray.jina.ai/fundamentals/document/) which wraps two [`chunks`](https://docarray.jina.ai/fundamentals/document/nested/) with the `image` and `text` modality.

Reviewer 3: ```suggestion
Each pair is stored in a [`Document`](https://docarray.jina.ai/fundamentals/document/) which wraps two [`chunks`](https://docarray.jina.ai/fundamentals/document/nested/) with `image` and `text` modality respectively.
```

You can push the resulting [`DocumentArray`](https://docarray.jina.ai/fundamentals/documentarray/) to the cloud using the [`.push`](https://docarray.jina.ai/api/docarray.array.document/?highlight=push#docarray.array.document.DocumentArray.push) method.

We use [fashion captioning dataset](https://github.com/xuewyang/Fashion_Captioning) as a sample dataset in this tutorial.

> You can get the description and image url from the dataset: 

Reviewer 3: ```suggestion
The following are examples of descriptions and image urls from the dataset: 
```

Reviewer 3: Maybe you can show the user how they can view these descriptions + urls themselves?

Author: Yes I can, but i think this will distract users and it is not the main topic of this tutorial

The following are examples of descriptions and image urls from the dataset: 

| Description                                                                                                                           | Image URL                                                                                                                                                           |
|---------------------------------------------------------------------------------------------------------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | subtly futuristic and edgy this liquid metal cuff bracelet is shaped from sculptural rectangular link                                 | [https://n.nordstrommedia.com/id/sr3/<br/>58d1a13f-b6b6-4e68-b2ff-3a3af47c422e.jpeg](https://n.nordstrommedia.com/id/sr3/58d1a13f-b6b6-4e68-b2ff-3a3af47c422e.jpeg) |

Author: copyright?

> | subtly futuristic and edgy this liquid metal cuff bracelet is shaped from sculptural rectangular link                                 | [https://n.nordstrommedia.com/id/sr3/<br/>58d1a13f-b6b6-4e68-b2ff-3a3af47c422e.jpeg](https://n.nordstrommedia.com/id/sr3/58d1a13f-b6b6-4e68-b2ff-3a3af47c422e.jpeg) |

Author: I also kept the original image urls since it is from the dataset

                ),
                Document(
                    uri='https://n.nordstrommedia.com/id/sr3/21e7a67c-0a54-4d09-a4a4-6a0e0840540b.jpeg',
                    modality='image',
                ),
            ],
        ),
        Document(
            chunks=[
                Document(
                    content='this shimmering tricot knit tote is traced with decorative whipstitching and diamond cut chain the two hallmark of the falabella line',
                    modality='text',
                ),
                Document(
                    uri='https://n.nordstrommedia.com/id/sr3/1d8dd635-6342-444d-a1d3-4f91a9cf222b.jpeg',
                    modality='image',
                ),
            ],
        ),
    ]
)
train_da.push('fashion-sample')
```

The full dataset has been converted to `clip-fashion-train-data` and `clip-fashion-eval-data` and pushed to the cloud.
You can directly use them in Finetuner.

## Start Finetuner

You may now create and run a fine-tuning job after login to Jina ecosystem.

```python
import finetuner

finetuner.login()
run = finetuner.fit(
    model='openai/clip-vit-base-patch32',
    run_name='clip-fashion',

>     train_data='clip-fashion-train-data',

Reviewer 2: yes, both are supported! it can be a string of docarray name, or a document array python instance

>     train_data='clip-fashion-train-data',

Reviewer: prettry cool

    eval_data='clip-fashion-eval-data',  # optional
    epochs=5,
    learning_rate=1e-5,
    loss='CLIPLoss',

>     cpu=False,

Reviewer 2: by default, finetuner use `cpu`, if you want to use `gpu`, you have to set `cpu=False`.

```

When the status is `FINISHED`, you can download the tuned model to your local machine.

```python
import finetuner

finetuner.login()
run = finetuner.get_run('clip-fashion')
run.save_artifact('clip-model')
```

You should now get a zip file containing the tuned model named `clip-fashion.zip` under the folder `clip-model`.

## Use the Model

After unzipping the model you get from the previous step, a folder with the following structure will be generated:

```text
.
└── clip-fashion/
    ├── config.yml
    ├── metadata.yml
    ├── metrics.yml
    └── models/
        ├── clip-text/
        │   ├── metadata.yml
        │   └── model.onnx
        ├── clip-vision/
        │   ├── metadata.yml
        │   └── model.onnx
        └── input-map.yml
```

Since the tuned model generated from Finetuner contains richer information such as metadata and config, we now transform it to simpler structure used by CLIP-as-service.

* First create a new folder named `clip-fashion-cas` or anything you like. This will be the storage of the models to use in CLIP-as-service.

* Second copy and move `clip-fashion/models/clip-text/model.onnx` to `clip-fashion-cas` and rename it to `textual.onnx`.

> * Similarly, copy and move `clip-fashion/models/clip-vision/model.onnx` to `clip-fashion-cas` and rename it to `visual.onnx`.

Reviewer 3: ```suggestion
* Similarly, copy the vision model `clip-fashion/models/clip-vision/model.onnx` into `clip-fashion-cas` and rename the model to `visual.onnx`.
```

This is the expected structure of `clip-fashion-cas`:

```text
.
└── clip-fashion-cas/
    ├── textual.onnx
    └── visual.onnx
```

> In order to use fine-tuned model, create a custom YAML file `finetuned_clip.yml` like below. Learn more about [Flow YAML configuration](https://docs.jina.ai/fundamentals/flow/yaml-spec/) and [`clip_server` YAML configuration](https://clip-as-service.jina.ai/user-guides/server/#yaml-config).

Reviewer 3: ```suggestion
In order to use the fine-tuned model, create a custom YAML file `finetuned_clip.yml` like below. Learn more about [Flow YAML configuration](https://docs.jina.ai/fundamentals/flow/yaml-spec/) and [`clip_server` YAML configuration](https://clip-as-service.jina.ai/user-guides/server/#yaml-config).
```
